src/main/java/com/aws/greengrass/builtin/services/telemetry/ComponentMetricIPCEventStreamAgent.java

```diff
@@ -0,0 +1,197 @@
+                logger.atDebug().kv(SERVICE_NAME, serviceName)
+                        .log("Publish Telemetry metric" + telemetryMetric.getName());
+                metricFactory.putMetricData(telemetryMetric);
+            });
+        }
+    }
+
+
+    // Creates telemetry metric object for given request metric
+    private Metric getTelemetryMetric(software.amazon.awssdk.aws.greengrass.model.Metric metric,
+                                      String metricNamespace) {
+        return Metric.builder()
+                .namespace(metricNamespace)
+                .name(metric.getName())
+                .unit(valueOfIgnoreCase(metric.getUnitAsString()))
+                .aggregation(TelemetryAggregation.Sum)
+                .value(metric.getValue())
+                .timestamp(Instant.now().toEpochMilli())
+                .build();
+    }
+
+    // Translate unit from metric request to telemetry unit
+    private TelemetryUnit valueOfIgnoreCase(String unitAsString) {
+        if (unitAsString == null || unitAsString.isEmpty()) {
+            throw new IllegalArgumentException("Invalid telemetry unit: Found null or empty value");
+        }
+
+        final String replacedString = String.join("", unitAsString.split(NON_ALPHANUMERIC_REGEX));
+        TelemetryUnit telemetryEnum = EnumUtils.getEnumIgnoreCase(TelemetryUnit.class, replacedString);
+
+        if (telemetryEnum == null || telemetryEnum.toString().isEmpty()) {
+            throw new IllegalArgumentException("Invalid telemetry unit: No matching TelemetryUnit type found");
+        }
+        return telemetryEnum;
+    }
+
+    // Validate metric request
+    // Check name, unit and value arguments
+    private void validateComponentMetricRequest(List<software.amazon.awssdk.aws.greengrass.model.Metric> metrics) {
+        if (Utils.isEmpty(metrics)) {
+            throw new IllegalArgumentException(
+                    String.format("Null or Empty list of metrics found in PutComponentMetricRequest"));
+        }
+        for (software.amazon.awssdk.aws.greengrass.model.Metric metric : metrics) {
+            if (Utils.isEmpty(metric.getName()) || metric.getName().getBytes(StandardCharsets.UTF_8).length > 32
+                    || Utils.isEmpty(metric.getUnitAsString())  || metric.getValue() < 0) {
+                throw new IllegalArgumentException(
+                        String.format("Invalid argument found in PutComponentMetricRequest"));
+            }
+        }
+    }
+
+    // Validate if request is authorized for given operation
+    // Also validate if serviceName is of format "aws.*"
+    private void doAuthorization(String opName, String serviceName) throws AuthorizationException {
+        if (authorizationHandler.isAuthorized(PUT_COMPONENT_METRIC_SERVICE_NAME,
+                Permission.builder().operation(opName).principal(serviceName).resource("*").build())
+                && SERVICE_NAME_REGEX.matcher(serviceName).find()) {
```

Comment:
nit: probably better to switch the order here so you check the name first and then check the authHandler. When a non aws service calls, it will log in the auth handler and customers could grant themselves the permissions and be confused as to why this is failing.